crates/types/src/replication/nodeset.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::hash::BuildHasherDefault;
+use std::iter::FusedIterator;
+
+use ahash::AHasher;
+use itertools::Itertools;
+use rand::prelude::*;
+
+use crate::PlainNodeId;
+
+// Why? Over 50% faster in iteration than HashSet and ~40% faster than default RandomState for
+// contains() and set intersection operations. Additionally, it's 300% faster when created from
+// iterators than HashSet with default RandomState.
+type IndexSet<T> = indexmap::IndexSet<T, BuildHasherDefault<AHasher>>;
+
+/// A type that represents a unique set of nodes. NodeSet maintains the order of the nodes in the
+/// set and provides efficient set operations. Note that the order across serialization and
+/// deserialization is only guaranteed if the underlying format maintains the order as well.
+#[derive(
+    serde::Serialize,
+    serde::Deserialize,
+    Clone,
+    Default,
+    Eq,
+    PartialEq,
+    derive_more::Index,
+    derive_more::IntoIterator,
+    derive_more::From,
+)]
+pub struct NodeSet(IndexSet<PlainNodeId>);
+
+impl NodeSet {
```

Comment:
If this will eventually replace the Placement in partition table we probably need to also have a `first() -> Option<&PlainNodeId>` (or similary) function to return the "leader" node in case of placement.

Or can be implemented as an Extension trait

Reply:
Yep, this is coming up in a subsequent PR. That will be in placement's newtype itself.